### ISSUE TITLE
Unable to login via OAuth2 - OmniAuth::Strategies::OAuth2::CallbackError

### ISSUE THREAD

**Issue by wilsonsilva** [hanami/hanami#650]
I was following [this tutorial](http://monterail.com/blog/2016/hanami-with-oauth/) to create [this example app](https://github.com/tobiaszwaszak/hanami_oauth_sample) which uses omniauth to login with Github. It was built with Hanami `v0.7.2` and it works absolutely fine.

But if I update Hanami to version `v0.8.0` it will no longer authenticate. This only happens if I run the app with `hanami server`. If I run the app with `rackup -p 2300` I'm able to authenticate with no problems.

This is the error I get when trying to authenticate:

```
Boot Error

Something went wrong while loading /Users/wilson/projects/hanami_oauth_sample/config.ru

OmniAuth::Strategies::OAuth2::CallbackError: OmniAuth::Strategies::OAuth2::CallbackError

/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/failure_endpoint.rb:25:in `raise_out!'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/failure_endpoint.rb:20:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/failure_endpoint.rb:12:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:479:in `fail!'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-oauth2-1.4.0/lib/omniauth/strategies/oauth2.rb:71:in `callback_phase'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:227:in `callback_call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:184:in `call!'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:164:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/builder.rb:63:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/warden-1.2.6/lib/warden/manager.rb:31:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:408:in `call_app!'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:362:in `callback_phase'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-oauth2-1.4.0/lib/omniauth/strategies/oauth2.rb:75:in `callback_phase'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:227:in `callback_call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:184:in `call!'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/strategy.rb:164:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/omniauth-1.3.1/lib/omniauth/builder.rb:63:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/warden-1.2.6/lib/warden/manager.rb:35:in `block in call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/warden-1.2.6/lib/warden/manager.rb:34:in `catch'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/warden-1.2.6/lib/warden/manager.rb:34:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:225:in `context'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:220:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/builder.rb:153:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/hanami-0.8.0/lib/hanami/middleware.rb:52:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/hanami-0.8.0/lib/hanami/application.rb:113:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/2.3.0/delegate.rb:83:in `method_missing'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/http_router-0.11.2/lib/http_router.rb:193:in `process_destination_path'
(eval):15:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/http_router-0.11.2/lib/http_router.rb:288:in `raw_call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/hanami-router-0.7.0/lib/hanami/routing/http_router.rb:156:in `raw_call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/http_router-0.11.2/lib/http_router.rb:142:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/hanami-router-0.7.0/lib/hanami/router.rb:969:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/methodoverride.rb:22:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/hanami-0.8.0/lib/hanami/assets/static.rb:49:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/builder.rb:153:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/hanami-0.8.0/lib/hanami/container.rb:47:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/shotgun-0.9.2/lib/shotgun/loader.rb:86:in `proceed_as_child'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/shotgun-0.9.2/lib/shotgun/loader.rb:31:in `call!'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/shotgun-0.9.2/lib/shotgun/loader.rb:18:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/hanami-0.8.0/lib/hanami/assets/static.rb:49:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/lint.rb:49:in `_call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/lint.rb:37:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/showexceptions.rb:24:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/commonlogger.rb:33:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/content_length.rb:15:in `call'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/handler/webrick.rb:88:in `service'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/2.3.0/webrick/httpserver.rb:140:in `service'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/2.3.0/webrick/httpserver.rb:96:in `run'
/Users/wilson/.rbenv/versions/2.3.1/lib/ruby/2.3.0/webrick/server.rb:296:in `block in start_thread'
```

and this is the server log:

```
[2016-10-06 20:08:41] INFO  WEBrick 1.3.1
[2016-10-06 20:08:41] INFO  ruby 2.3.1 (2016-04-26) [x86_64-darwin15]
[2016-10-06 20:08:41] INFO  WEBrick::HTTPServer#start: pid=6238 port=2300
::1 - - [06/Oct/2016:20:08:44 +0100] "GET /auth/signout HTTP/1.1" 302 - 0.0477
::1 - - [06/Oct/2016:20:08:44 +0100] "GET / HTTP/1.1" 200 - 0.0497
I, [2016-10-06T20:08:45.804160 #6253]  INFO -- omniauth: (github) Request phase initiated.
::1 - - [06/Oct/2016:20:08:45 +0100] "GET /auth/github HTTP/1.1" 302 252 0.0349
I, [2016-10-06T20:08:46.371948 #6254]  INFO -- omniauth: (github) Callback phase initiated.
I, [2016-10-06T20:08:48.222670 #6254]  INFO -- omniauth: (github) Callback phase initiated.
E, [2016-10-06T20:08:48.223010 #6254] ERROR -- omniauth: (github) Authentication failure! csrf_detected: OmniAuth::Strategies::OAuth2::CallbackError, csrf_detected | CSRF detected
E, [2016-10-06T20:08:48.223224 #6254] ERROR -- omniauth: (github) Authentication failure! invalid_credentials: OmniAuth::Strategies::OAuth2::CallbackError, csrf_detected | CSRF detected
E, [2016-10-06T20:08:48.223642 #6254] ERROR -- omniauth: (github) Authentication failure! invalid_credentials: OmniAuth::Strategies::OAuth2::CallbackError, csrf_detected | CSRF detected
```

It seems that running the app with `hanami server` causes the callback to be executed twice.
## Steps to reproduce:

Setup project
- `git clone git@github.com:tobiaszwaszak/hanami_oauth_sample.git`
- `cd hanami_oauth_sample`
- `bundle install`
- Register a [new application](https://github.com/settings/developers) on Github
- On Github, set **Homepage URL** to `http://localhost:2300`
- On Github, set **Authorization callback URL** to `http://localhost:2300/auth/github/callback`
- Copy the **Client ID** and **Client Secret** of the new app to [env.development](https://github.com/tobiaszwaszak/hanami_oauth_sample/blob/master/.env.development#L5-L6)

Verifying that it works on Hanami `v0.7.2`
- `bundle exec hanami server`
- Open `http://localhost:2300` and login into Github. It **should work**.

Verifying that it does not work on Hanami `v0.8` with the command `hanami serve`
- Logout and stop the hanami server
- Now update Hanami to version `0.8`
- `bundle install`
- `bundle exec hanami server`
- Open `http://localhost:2300` and login into Github. It **should not work**.

Verifying that it works on Hanami `v0.8` with the command `rackup -p 2300`
- Close hanami server
- `rackup -p 2300`
- Open `http://localhost:2300` and login into Github. It **should work**.

**Comment by mereghost** [comment]
Maybe it's related to #643 ?

**Comment by tobiaszwaszak** [comment]
Yes, its related to #643 
I replaced hanami gem with this PR and it works fine

**Comment by jodosha** [comment]
@wilsonsilva Thanks for reporting this. #643 Is fixed in a soon-to-be-merged branch. I'll let you to know when. Thanks! /cc @mereghost @tobiaszwaszak

**Comment by jodosha** [comment]
Fixed by #657
